File: cpp/078_Subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        std::sort(nums.begin(),nums.end());
        vector<vector<int>> ret;
        vector<int> path;
        backtrace(ret,path,nums,0);
        return ret;
    }   
    
private:
    void backtrace(vector<vector<int>>& ret, vector<int>& path, vector<int>& nums,int index){
        ret.push_back(path);
        for(int i = index; i < nums.size(); ++i){
            path.push_back(nums[i]);
            backtrace(ret,path,nums,i+1);
            path.pop_back();
        }
    }
};

File: js/078_Subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    nums.sort( function(a, b){ return a - b; } );
    var ret = [];
    backtrace(ret,[],nums,0);
    return ret;
};

var backtrace = function(ret,path,nums,index){
    ret.push(path.concat());
    for(var i = index; i < nums.length; ++i){
        path.push(nums[i]);
        backtrace(ret,path,nums,i+1);
        path.pop();
    }
};